remote-api/src/main/java/org/alfresco/rest/api/model/ActionParameterConstraint.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * #%L
+ * Alfresco Remote API
+ * %%
+ * Copyright (C) 2005 - 2021 Alfresco Software Limited
+ * %%
+ * This file is part of the Alfresco software.
+ * If the software was purchased under a paid Alfresco license, the terms of
+ * the paid license agreement will prevail.  Otherwise, the software is
+ * provided under the following open source license terms:
+ *
+ * Alfresco is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Alfresco is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Alfresco. If not, see <http://www.gnu.org/licenses/>.
+ * #L%
+ */
+
+package org.alfresco.rest.api.model;
+
+import java.util.List;
+import java.util.Objects;
+
+import org.alfresco.service.Experimental;
+
+/**
+ * Representation of action parameter constraint.
+ *
+ * @author mpichura
+ */
+@Experimental
+public class ActionParameterConstraint
+{
+
+    /**
+     * Constraint name.
+     */
+    private String constraintName;
+    /**
+     * List of objects representing constraint values along with additional data
+     */
+    private List<ConstraintData> constraintValues;
+
+    public List<ConstraintData> getConstraintValues()
+    {
+        return constraintValues;
+    }
+
+    public void setConstraintValues(List<ConstraintData> constraintValues)
+    {
+        this.constraintValues = constraintValues;
+    }
+
+    public String getConstraintName()
+    {
+        return constraintName;
+    }
+
+    public void setConstraintName(String constraintName)
+    {
+        this.constraintName = constraintName;
+    }
+
+    public static class ConstraintData
+    {
+        public ConstraintData(final String value, final String label)
+        {
+            this.value = value;
+            this.label = label;
+        }
+
+        public ConstraintData(final String value, final String label, final boolean isNode)
+        {
+            this.value = value;
+            this.label = label;
+            this.isNode = isNode;
+        }
+
+        private String value;
+        private String label;
+        private Boolean isNode;
```

Comment:
I don't know (and can't guess) how `isNode` is meant to work.  Please can we have javadoc for all three of these fields and the class itself?

Reply:
I still don't understand this. How can `value` be the "Actual constraint value"?  Do you mean the parameter value?

Reply:
This class represents constraints for parameter values. So with sample action definition of:
```
"entry": {
                    "applicableTypes": [],
                    "parameterDefinitions": [
                        {
                            "name": "aspect-name",
                            "type": "d:qname",
                            "multiValued": false,
                            "mandatory": true,
                            "displayLabel": "Aspect",
                            "parameterConstraintName": "ac-aspects"
                        }
                    ],
                    "name": "add-features",
                    "trackStatus": false,
                    "description": "This will add an aspect to the matched item.",
                    "adhocPropertiesAllowed": true,
                    "id": "add-features",
                    "title": "Add aspect"
                }
```
We need to create an action with name `"add-features"` and exactly one parameter entry where the key of that entry is `"aspect-name"` and possible values are coming from the `value` field in list of `ActionParameterConstraint.ConstraintData` 
So this `value` can be treated as "possible parameter value" or parameter constraint. The `value` filed is there because we also provide display `label` and `isNode` indicator (for some of the constraints).
This class could possibly be avoided and the list of constraints could be List<String> which would be constraints / possible parameter values.
I am  happy to take some advice in here as I also see that this is a bit misleading.